### ISSUE TITLE
Some battle results share UUIDs, thus recognized as duplicates & rejected by stat.ink

### ISSUE THREAD

**Issue by wutSalty** [frozenpandaman/s3s#29]
I decided to upload all of my locally exported data (using -i) to stat.ink and while a majority of the matches seem to have gone through properly, I noticed a few of them were coming up as already been uploaded one after another. In one instance, it did not seem to upload any match results at all. The command prompt does not seem to raise any other errors in regards to those duplicated entries.

Because of the Splatfest, it was hard for me to exactly cross-reference the matches I played recently vs the matches that have successfully been uploaded to stat.ink because it does not support Splatfest battles yet. However, I played 5 matches of regular Turf Wars after the Splatfest. Upon uploading using -r, it only detected 1 of these matches, and showed the other 4 as being a duplicate.

I am currently running s3s on Windows 10 Version 10.0.19044.2006 with Python 3.10.7 installed through the Windows Command Prompt, without Administrator Access.

I have attached the screenshots of some of the uploads that return with duplicated entries, as well as the JSON files generated with the local export. Screenshots of my recent Turf matches from Splatnet along with the local exported data is also attached.
(Uploaded 2 separate zip files since 1 seemed to exceed Github's limit)
[exports(1).zip](https://github.com/frozenpandaman/s3s/files/9647052/exports.1.zip)
[exports(2).zip](https://github.com/frozenpandaman/s3s/files/9647060/exports.2.zip)

**Comment by Rexogamer** [comment]
I've also encountered this issue - here's an example: 

```
> python s3s.py -nsr
Number of recent battles to upload (0-50)? 12
Pulling data from online...
Validating your tokens... done.

Battle already uploaded - https://stat.ink/@Rexogamer/spl3/16c2cdfc-e52e-4615-94c7-3a6b70ef3fc0
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/d37675a1-aa9b-4cf7-801a-2fc48b21bffa
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/d37675a1-aa9b-4cf7-801a-2fc48b21bffa
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/d37675a1-aa9b-4cf7-801a-2fc48b21bffa
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/8d361408-654b-4ed4-b9f7-602939a93923
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/cf97b04e-6564-4cef-856b-6c2f2417b760
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/cf97b04e-6564-4cef-856b-6c2f2417b760
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/cf97b04e-6564-4cef-856b-6c2f2417b760
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/3749c0a4-bc25-4f57-b06a-71267d353522
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/7864172d-1775-432f-8669-1b315698a65e
Battle already uploaded - https://stat.ink/@Rexogamer/spl3/7864172d-1775-432f-8669-1b315698a65e
Battle uploaded to https://stat.ink/@Rexogamer/spl3/cac308c3-c0e8-41e3-91d2-54f0c418fc92
```

Here are the results and overview files:
[results.zip](https://github.com/frozenpandaman/s3s/files/9649174/results.zip)

**Comment by frozenpandaman** [comment]
This seems to be because some battles share the same UUIDs as others:
![image](https://user-images.githubusercontent.com/7245874/192401441-12fc945f-034f-4ae9-902c-92f8cb144b77.png)

The battles are indeed different (different players, results, etc.) so I have no idea why why they would be given the same UUID part of their Base64-encoded `id`. We may need to change the `uuid` strings we send to stat.ink to avoid this.

**Comment by frozenpandaman** [comment]
I'm guessing that the SplatNet 3 UUID – included at the end of each battle's Base64-encoded `id` – is unique to each specific lobby/group, i.e. if you choose to continue playing in Turf War, and the "host" console/user (not shown to players who this is) decides to continue as well, the following battle record will have the same UUID, and just a different time prefix. This also explains why ~1–2 players seemed to always be shared between the battles (and why it didn't seem to happen for Anarchy battles, since I think a new lobby/"room" is created every battle, unlike TW).

Talked with Hina and we'll be changing the payload `uuid` parameter we send stat.ink – now a UUID v5 based on a unique s3s UUID namespace and the `<YYYYMMDD>T<HHMMSS>_<uuid>` part of the SplatNet 3 battle `id`. Old `uuid`s will still be checked to prevent (re)uploading duplicate battles, unless the new `force_uploads` [config key](https://github.com/frozenpandaman/s3s/wiki/config-keys) is added.

**Comment by frozenpandaman** [comment]
Fixed in v0.1.4.

Only Nintendo would make their UUIDs non-unique… 🙃 (ok, to be fair, they still kind of are, depending on your definitions & expectations… just not per battle – per hidden/internal 'lobby' mechanism…)

**Comment by Adrthegamedev** [comment]
ruh roh, it does Not appear that the duplication check worked; ran `python3 s3s.py -r` and now i have a bunch of duplicate battles on my stat.ink page now, with no way of deleting them :stress:

**Comment by Adrthegamedev** [comment]
(@frozenpandaman btw, i keep forgetting that github may be dumb and not notify people)

**Comment by frozenpandaman** [comment]
@Adrthegamedev Oh oops, it'll work for old battles with `-i` but not when you re-run it with recent `-r` ones… 🫠 Sorry about that.

You can delete them by sending DELETE requests using a REST client like Postman: https://github.com/fetus-hina/stat.ink/wiki/Spl3-API:-Delete-v3-battle

If that's too complicated, or too many to do one-by-one, let me know and I can post a Gist for a script that I've been using to mass delete mine for testing.

(and don't worry, as the owner I get notified about everything in this repo haha)

**Comment by Adrthegamedev** [comment]
yeah, that would be handy; would give me a chance to chronologically order my stat.ink splatoon 3 battles rather than just slowly deleting them via postman one by one

**Comment by wutSalty** [comment]
Just ran through all of my past exports and while the matches played just now seem to have been uploaded fine, the matches from previous days still show as an "already-uploaded battle". Would I have to delete all of the older matches from stat.ink first before uploading them again?

**Comment by frozenpandaman** [comment]
>Oh I was _just_ testing whether the re-used IDs were because of continuing in the same Turf lobby. 

@wutSalty Cool! It's kind of just for my own curiosity at this point, but if you can test/confirm that, that'd be cool :D

**Comment by frozenpandaman** [comment]
> the matches from previous days still show as an "already-uploaded battle". Would I have to delete all of the older matches from stat.ink

@wutSalty Why do you want to delete & re-upload them? There's no need, unless you want to fix the chronological order stuff (but this is already an issue since there's no Splatfest support yet, so… there'll already be a 2-day gap…) Hina will fix the kill counts mentioned in #30 soon.

**Comment by wutSalty** [comment]
I'm trying to upload the previous battles that were skipped the first time. But running -i with those exports now seem to show that the battle was already uploaded (when they haven't yet)

**Comment by wutSalty** [comment]
Would that still skip over the entries that have already been uploaded? Or would I have to go through and manually delete them afterwards.

**Comment by frozenpandaman** [comment]
@wutSalty It will just upload _everything_ again (sorry – it's in alpha/beta after all 🙃). Don't think there's any way to fix this – just use the mass deletion script I'm about to post in a minute to clean up beforehand and start anew, imo.

**Comment by wutSalty** [comment]
Ah that's fine, early adopter perks lol. Yeah I'll wait for the script so that I can wipe everything from stat.ink first then I'll go through and upload them again. All goods.

**Comment by Adrthegamedev** [comment]
sideeffect of this fix; `-r` duplicate battle checking doesn't work, even with the new uuids after wiping all stat.ink battles;
it says `Previously-unuploaded battles detected. Uploading now...` and then a flood of `Battle already uploaded` afterwards

**Comment by frozenpandaman** [comment]
There's like 10 conversations going on here at once but. Following up on an old comment…

>it'll work for old battles with `-i` but not when you re-run it with recent `-r` ones… 

@Adrthegamedev 8dcff70 fixes it in `-r` now too – thanks for the heads up there. Once you clean up your dupes, if you're able to try it again & confirm it works, that'd be cool!

**Comment by Adrthegamedev** [comment]
already cleaned up all my dupes; it does indeed correctly detect now that all that's not been uploaded is the, like, 50+ splatfest battles i've played

**Comment by frozenpandaman** [comment]
>I can post a Gist for a script that I've been using to mass delete mine for testing

# `statinkmassdel.py`

https://gist.github.com/frozenpandaman/d531a9a45bf13fef5ce8fbccc3fcc69c

I do mine a page at a time:

1. Copy the HTML source for the table on your user profile.
2. Replace all instances of `/@yourusername` with `https://stat.ink/@yourusername`
3. Extract all the links using something like https://www.convertcsv.com/url-extractor.htm
4. Wrap them in `"`…`",` using your fancy multi-cursor text editor tool or whatever
5. Paste them into the script, fill in your stat.ink API key, and run it

cc @Peri0dic @Adrthegamedev

**Comment by Adrthegamedev** [comment]
> @Adrthegamedev If you have actually deleted all battles from stat.ink it will not give you any "Battle already uploaded" messages.

well, i said that at a point where I,
\- A. deleted all battles
and 
\- B. reuploaded everything correctly via `-i`ing my excessive backups chronologically

**Comment by wutSalty** [comment]
> > Oh I was _just_ testing whether the re-used IDs were because of continuing in the same Turf lobby.
> 
> @wutSalty Cool! It's kind of just for my own curiosity at this point, but if you can test/confirm that, that'd be cool :D

Quickly had to figure out how to get Notepad++ to play nice with results.json and neaten the file so I could actually read it. But I can confirm that the "ID" seems to be a lobby/host ID. I cross-referenced results.json to the matches I played and for games where I pressed "Keep Going", I would get the same ID. For lobbies where I quit and found a different lobby, it would appear with a different ID. Funnily enough, there was one instance where I joined matching too quickly and found myself in the same lobby that I just left (which did give the same ID).

Attached is just a text file showing the action I took with each match (either re-joined or continued) and the ID that I found for that battle.
[joined_vs_continued.txt](https://github.com/frozenpandaman/s3s/files/9656673/joined_vs_continued.txt)
